### PR TITLE
make sure failed neighbor list searches trigger exhaustive searches

### LINKED ISSUE #1827: Neighbor Lists Never Populated
It appears that cell neighbor lists are not being populated in the current version of OpenMC -- i.e., the neighbor list `push_back()` function is never called. From what I can tell, this regression was introduced in #1784 when recursive calls were eliminated from the cell finding process.

The issue stems from a change made in find_cell_inner() of geometry.cpp. The old version basically had the following logic:

```
if( neighbor_list )
    ... check neighbor list cells
else
    ... do exhaustive search
```

The new version was changed to do

```
if( neighbor_list )
    ... check neighbor list cells

do exhaustive search
```

This means that the `find_cell_inner()` function will always find the cell, even if a neighbor list is passed and the cell is not in the neighbor list. This is a problem, due to the way find_cell_inner is called:

```C++
// Search for the particle in that cell's neighbor list.  Return if we
// found the particle.
bool found = find_cell_inner(p, &c.neighbors_);
if (found) return found;

// The particle could not be found in the neighbor list.  Try searching all
// cells in this universe, and update the neighbor list if we find a new
// neighboring cell.
found = find_cell_inner(p, nullptr);
if (found) c.neighbors_.push_back(p.coord_[coord_lvl].cell);
return found;
```

This regression could potentially have a significant performance impact on a lot of problems. It may be worth doing some analysis to see if a new release with a patch is worthwhile, as this PR went in right before 12.1.

### PR DESCRIPTION
**Pull request by gridley**
Fixes #1827 

So, I slipped up slightly when removing recursion. I had assumed that if we try to iterate through neighbor lists, it will always touch the variable `i_cell` in find_cell_inner. Well, that's actually not true, since it's possible that the neighbor lists are empty, which in that case, i_cell will never be affected. As a result of this, we were never forming neighbor lists.

The fix to this is simple. If we're attempting a cell search based on neighbor lists, we should always return false if we failed to find the right cell in the neighbor list. This will cause an exhaustive search to happen, the result of which will be appended to the neighbor list.

Here is a simple 5x5 checkerboard input done directly using surfaces, so that the impact of neighbor lists is fairly high. On my computer, I saw the tracking rate go from around 190 kp/s to around 240 kp/s.

```
<?xml version='1.0' encoding='utf-8'?>
<geometry>
  <!-- 25 cells in total -->
  <cell material="1" id="1"  region="1 -5 3 -9" universe="0" />
  <cell material="2" id="2"  region="5 -6 3 -9" universe="0" />
  <cell material="1" id="3"  region="6 -7 3 -9" universe="0" />
  <cell material="2" id="4"  region="7 -8 3 -9" universe="0" />
  <cell material="1" id="5"  region="8 -2 3 -9" universe="0" />
  <cell material="2" id="6"  region="1 -5 9 -10" universe="0" />
  <cell material="1" id="7"  region="5 -6 9 -10" universe="0" />
  <cell material="2" id="8"  region="6 -7 9 -10" universe="0" />
  <cell material="1" id="9"  region="7 -8 9 -10" universe="0" />
  <cell material="2" id="10" region="8 -2 9 -10" universe="0" />
  <cell material="1" id="11" region="1 -5 10 -11" universe="0" />
  <cell material="2" id="12" region="5 -6 10 -11" universe="0" />
  <cell material="1" id="13" region="6 -7 10 -11" universe="0" />
  <cell material="2" id="14" region="7 -8 10 -11" universe="0" />
  <cell material="1" id="15" region="8 -2 10 -11" universe="0" />
  <cell material="2" id="16" region="1 -5 11 -12" universe="0" />
  <cell material="1" id="17" region="5 -6 11 -12" universe="0" />
  <cell material="2" id="18" region="6 -7 11 -12" universe="0" />
  <cell material="1" id="19" region="7 -8 11 -12" universe="0" />
  <cell material="2" id="20" region="8 -2 11 -12" universe="0" />
  <cell material="1" id="21" region="1 -5 12 -4" universe="0" />
  <cell material="2" id="22" region="5 -6 12 -4" universe="0" />
  <cell material="1" id="23" region="6 -7 12 -4" universe="0" />
  <cell material="2" id="24" region="7 -8 12 -4" universe="0" />
  <cell material="1" id="25" region="8 -2 12 -4" universe="0" />

  <surface boundary="reflective" coeffs="0" id="1" type="x-plane" />
  <surface boundary="reflective" coeffs="10" id="2" type="x-plane" />
  <surface boundary="reflective" coeffs="0" id="3" type="y-plane" />
  <surface boundary="reflective" coeffs="10" id="4" type="y-plane" />

  <surface coeffs="2" id="5"  type="x-plane" />
  <surface coeffs="4" id="6"  type="x-plane" />
  <surface coeffs="6" id="7"  type="x-plane" />
  <surface coeffs="8" id="8"  type="x-plane" />

  <surface coeffs="2" id="9"  type="y-plane" />
  <surface coeffs="4" id="10" type="y-plane" />
  <surface coeffs="6" id="11" type="y-plane" />
  <surface coeffs="8" id="12" type="y-plane" />
</geometry>
<?xml version='1.0' encoding='utf-8'?>
<materials>
  <material depletable="true" id="1" name="Fuel 3.1%" temperature="300">
    <density units="g/cc" value="10.30166" />
    <nuclide ao="1.9992419999999977" name="O16" />
    <nuclide ao="0.0007579999999999991" name="O17" />
    <nuclide ao="0.00028071399816198897" name="U234" />
    <nuclide ao="0.03140630756586438" name="U235" />
    <nuclide ao="0.9681691224623687" name="U238" />
    <nuclide ao="0.0001438559736051798" name="U236" />
  </material>
  <material id="2" name="Water" temperature="300">
    <density units="g/cc" value="0.7405820675158279" />
    <nuclide ao="0.0003217866044341838" name="B10" />
    <nuclide ao="0.001301758322075321" name="B11" />
    <nuclide ao="1.9964419358487546" name="H1" />
    <nuclide ao="0.00031097429822629077" name="H2" />
    <nuclide ao="0.9979980703970175" name="O16" />
    <nuclide ao="0.00037838467647285283" name="O17" />
  </material>
</materials>
<?xml version='1.0' encoding='utf-8'?>
<plots>
  <plot basis="xy" color_by="material" id="1" type="slice">
    <origin>5 5 0.5</origin>
    <width>10 10</width>
    <pixels>800 800</pixels>
  </plot>
</plots>
<?xml version='1.0' encoding='utf-8'?>
<settings>
  <run_mode>eigenvalue</run_mode>
  <particles>100000</particles>
  <batches>5</batches>
  <inactive>3</inactive>
  <source strength="1.0">
    <space type="box">
      <parameters>0 10 0 10 1 1</parameters>
    </space>
  </source>
</settings>

```

### CODE REVIEW

File: src/geometry.cpp [hunk 1]
```diff
@@ -91,11 +91,25 @@ find_cell_inner(Particle& p, const NeighborList* neighbor_list)
...
   for (;;++p.n_coord_) {
-    // If neighbor lists did not find a cell, must do exhaustive search
+    // If we did not attempt to use neighbor lists, i_cell is still C_NONE.  In
+    // that case, we should now do an exhaustive search to find the right value
+    // of i_cell.
+    //
+    // Alternatively, neighbor list searches could have succeeded, but we found
+    // that the fill of the neighbor cell was another universe. As such, in the
+    // code below this conditional, we set i_cell back to C_NONE to indicate
+    // that.
```
Comment (paulromano) [inline src/geometry.cpp:112]: One other thing I noticed when looking at the refactor of this code is that we should be resetting the `found` variable to false at the end of the `for (;;++p.n_coord_)` loop. In the recursive version, the recursive call to `find_cell_inner` resulted in a new `found` stack variable initialized to false, but in our loop version, it doesn't get reset.

**Review by paulromano — CHANGES_REQUESTED** [review]
Thanks for implementing a fix so quickly @gridley!

File: src/geometry.cpp (hunk 1 above)
Comment (gridley) [reply src/geometry.cpp:112]: Ah yes, good observation. It makes me think we should have some regression tests that we know fail in a certain way. I think we would have gone from returning a helpful error message to a segfault as a result of this.

Comment (paulromano) [reply src/geometry.cpp:112]: Yes, it's pretty easy to demonstrate actually. If you have an invalid geometry where part of a universe is missing, `find_cell` will result in a segfault. Here's a short snippet that shows this behavior:
```Python
import openmc
import openmc.lib

m = openmc.Material()
m.add_nuclide('U235', 1.0)
m.set_density('g/cm3', 5.0)
openmc.Materials([m]).export_to_xml()

plane = openmc.XPlane()
lower_cell = openmc.Cell(fill=m, region=-plane)
# Note there's no cell filling +plane
lower_univ = openmc.Universe(cells=[lower_cell])

cyl = openmc.ZCylinder(boundary_type='reflective')
cell_outer = openmc.Cell(fill=lower_univ, region=-cyl)
geom = openmc.Geometry([cell_outer])
geom.export_to_xml()

settings = openmc.Settings()
settings.batches = 10
settings.inactive = 5
settings.particles = 1000
settings.export_to_xml()

openmc.lib.init()
openmc.lib.find_cell((0.1, 0., 0.))
```

File: src/geometry.cpp [hunk 2]
```diff
@@ -91,11 +91,25 @@ find_cell_inner(Particle& p, const NeighborList* neighbor_list)
         break;
       }
     }
+
+    // If we're attempting a neighbor list search and fail, we
+    // now know we should return false. This will trigger an
+    // exhaustive search from neighbor_list_find_cell and make
+    // the result from that be appended to the neighbor list.
+    if (!found)
+      return false;
```
Comment (jtramm) [inline src/geometry.cpp:100]: A few lines down (at line 140) we have a similar `!found` check. That one returns `found`, this one returns `false`. Might make things slightly easier to read if both logic checks were identical.

File: src/geometry.cpp (hunk 1 above)
Comment (gridley) [reply src/geometry.cpp:112]: OK, next commit should fix this.

File: src/geometry.cpp (hunk 2 above)
Comment (gridley) [reply src/geometry.cpp:100]: True. I think it's more expressive to put `return found` so, I've changed that.

**Comment by gridley** [comment]
OK, the latest commit should address all the concerns.

**Review by paulromano — APPROVED** [review]

**Pull request merged**